File: app/src/main/java/eu/warble/pjapp/ui/student_info/fees/StudentFeesPresenter.java
package eu.warble.pjapp.ui.student_info.fees;


import android.os.Bundle;
import android.support.annotation.Nullable;

import java.util.List;

import eu.warble.pjapp.data.StudentDataRepository;
import eu.warble.pjapp.data.StudentDataSource;
import eu.warble.pjapp.data.local.StudentLocalDataSource;
import eu.warble.pjapp.data.model.OplatyItem;
import eu.warble.pjapp.data.model.PlatnosciItem;
import eu.warble.pjapp.data.model.Student;
import eu.warble.pjapp.data.remote.PjatkAPI;
import eu.warble.pjapp.ui.base.BaseFragmentPresenter;
import eu.warble.pjapp.util.AppExecutors;
import eu.warble.pjapp.util.CredentialsManager;

public class StudentFeesPresenter extends BaseFragmentPresenter<StudentFeesFragment> {

    public StudentFeesPresenter(StudentFeesFragment fragment) {
        super(fragment);
    }

    @Override
    protected void initPresenter(@Nullable Bundle savedInstanceState) {
        loadData();
    }

    private void loadData(){
        AppExecutors executors = new AppExecutors();
        StudentDataRepository repository = StudentDataRepository.getInstance(
                PjatkAPI.getInstance(CredentialsManager.getInstance().getCredentials(fragment.context), executors),
                StudentLocalDataSource.getInstance(executors)
        );
        repository.getStudentData(new StudentDataSource.LoadStudentDataCallback() {
            @Override
            public void onDataLoaded(Student studentData) {
                fillViews(studentData);
            }

            @Override
            public void onDataNotAvailable(String error) {
                fragment.showError(error);
            }
        });
    }

    private void fillViews(Student studentData){
        fillBalance(studentData);
        fillPayments(studentData);
        fillReceivables(studentData);
        fragment.setAccount(studentData.getKontoWplat());
    }

    private void fillBalance(Student studentData) {
        fragment.setBalance(String.valueOf(studentData.getSaldo()));
        fragment.setBalancePayments(String.valueOf(studentData.getKwotaWplat()));
        fragment.setBalanceReceivables(String.valueOf(studentData.getKwotaNaleznosci()));
        fragment.setBalanceWithdrawals(String.valueOf(studentData.getKwotaWyplat()));
    }

    private void fillPayments(Student studentData){
        List<PlatnosciItem> platnosci = studentData.getPlatnosci();
        if (!platnosci.isEmpty()){
            fragment.setPaymentsDate1(platnosci.get(platnosci.size()-1).getDataWplaty());
            fragment.setPaymentsDate1Amount(platnosci.get(platnosci.size()-1).getKwota() + "");
            if (platnosci.size() >=2) {
                fragment.setPaymentsDate2(platnosci.get(platnosci.size() - 2).getDataWplaty());
                fragment.setPaymentsDate2Amount(platnosci.get(platnosci.size() - 2).getKwota() + "");
            }
        }
    }

    private void fillReceivables(Student studentData){
        List<OplatyItem> oplaty = studentData.getOplaty();
        if (!oplaty.isEmpty()){
            fragment.setReceivables_item1(oplaty.get(oplaty.size()-1).getNazwa());
            fragment.setReceivables_item1_date(oplaty.get(oplaty.size()-1).getTerminPlatnosci());
            fragment.setReceivables_item1_amount(oplaty.get(oplaty.size()-1).getKwota() + "");
            if (oplaty.size() >= 2) {
                fragment.setReceivables_item2(oplaty.get(oplaty.size()-2).getNazwa());
                fragment.setReceivables_item2_date(oplaty.get(oplaty.size()-2).getTerminPlatnosci());
                fragment.setReceivables_item2_amount(oplaty.get(oplaty.size()-2).getKwota() + "");
            }
        }
    }
}

File: app/src/main/java/eu/warble/pjappkotlin/mvp/studentinfo/fees/StudentFeesPresenter.kt
package eu.warble.pjappkotlin.mvp.studentinfo.fees

import android.content.Context
import eu.warble.pjappkotlin.data.StudentDataRepository
import eu.warble.pjappkotlin.data.StudentDataSource
import eu.warble.pjappkotlin.data.model.Student

class StudentFeesPresenter(
        private val studentDataRepository: StudentDataRepository?,
        val view: StudentFeesContract.View,
        private val appContext: Context
) : StudentFeesContract.Presenter {

    override fun start() {
        studentDataRepository?.getStudentData(appContext, object : StudentDataSource.LoadStudentDataCallback {
            override fun onDataLoaded(studentData: Student) {
                fillViews(studentData)
            }

            override fun onDataNotAvailable(error: String) {
                view.showError(error)
            }
        })
    }

    private fun fillViews(studentData: Student) {
        with(view) {
            //balance card
            setBalance(studentData.saldo.toString())
            setBalancePayments(studentData.kwotaWplat.toString())
            setBalanceReceivables(studentData.kwotaNaleznosci.toString())
            setBalanceWithdrawals(studentData.kwotaWyplat.toString())
            //payments card
            val platnosci = studentData.platnosci
            if (platnosci?.isNotEmpty() == true) {
                setPaymentsDate1(platnosci[platnosci.size - 1].dataWplaty)
                setPaymentsDate1Amount(platnosci[platnosci.size - 1].kwota.toString())
                if (platnosci.size >= 2) {
                    setPaymentsDate2(platnosci[platnosci.size - 2].dataWplaty)
                    setPaymentsDate2Amount(platnosci[platnosci.size - 2].kwota.toString())
                }
            }
            //receivables card
            val oplaty = studentData.oplaty
            if (oplaty?.isNotEmpty() == true) {
                setReceivablesItem1(oplaty[oplaty.size - 1].nazwa)
                setReceivablesItem1Date(oplaty[oplaty.size - 1].terminPlatnosci)
                setReceivablesItem1Amount(oplaty[oplaty.size - 1].kwota.toString())
                if (oplaty.size >= 2) {
                    setReceivablesItem2(oplaty[oplaty.size - 2].nazwa)
                    setReceivablesItem2Date(oplaty[oplaty.size - 2].terminPlatnosci)
                    setReceivablesItem2Amount(oplaty[oplaty.size - 2].kwota.toString())
                }
            }
            //balance card
            setAccount(studentData.kontoWplat)
        }
    }

}